gpiozero/mixins.py

```diff
@@ -270,13 +280,119 @@ def _fire_events(self):
+        self._is_held = False
+        self._hold_time = 1
+        self._hold_repeat = False
+        self._hold_thread = RepeatingThread(self)
+
+    def close(self):
+        if self._hold_thread:
+            self._hold_thread.stop()
+            self._hold_thread = None
+        try:
+            super(RepeatingMixin, self).close()
+        except AttributeError:
+            pass
+
+    def _fire_activated(self):
+        super(RepeatingMixin, self)._fire_activated()
+        self._hold_thread.holding.set()
+
+    def _fire_deactivated(self):
+        self._hold_thread.holding.clear()
+        super(RepeatingMixin, self)._fire_deactivated()
+
+    def _fire_held(self):
+        if self.when_held:
+            self.when_held()
+
+    @property
+    def when_held(self):
+        """
+        The function to run when the device has remained active for
+        :attr:`hold_time` seconds.
+
+        This can be set to a function which accepts no (mandatory) parameters,
+        or a Python function which accepts a single mandatory parameter (with
+        as many optional parameters as you like). If the function accepts a
+        single mandatory parameter, the device that activated will be passed
+        as that parameter.
+
+        Set this property to ``None`` (the default) to disable the event.
+        """
+        return self._when_held
+
+    @when_held.setter
+    def when_held(self, value):
+        self._when_held = self._wrap_callback(value)
+
+    @property
+    def hold_time(self):
+        """
+        The length of time (in seconds) to wait after the device is activated,
+        until executing the :attr:`when_held` handler. If :attr:`hold_repeat`
+        is True, this is also the length of time between invocations of
+        :attr:`when_held`.
+        """
+        return self._hold_time
+
+    @hold_time.setter
+    def hold_time(self, value):
+        self._hold_time = value
+
+    @property
+    def hold_repeat(self):
+        """
+        If ``True``, :attr:`when_held` will be executed repeatedly with
+        :attr:`hold_time` seconds between each invocation.
+        """
+        return self._hold_repeat
+
+    @hold_repeat.setter
+    def hold_repeat(self, value):
+        self._hold_repeat = bool(value)
+
+    @property
+    def is_held(self):
+        """
+        When ``True``, the device has been active for at least
+        :attr:`hold_time` seconds.
+        """
+        return self._is_held
+
+
+class RepeatingThread(GPIOThread):
+    """
+    Extends :class:`GPIOThread`. Provides a background thread that repeatedly
+    fires the :attr:`RepeatingEventsMixin.when_held` event as long as the
+    owning device is active.
+    """
+    def __init__(self, parent):
+        super(RepeatingThread, self).__init__(target=self.held, args=(parent,))
+        self.holding = Event()
+        self.start()
+
+    def held(self, parent):
+        while not self.stopping.wait(0):
+            if self.holding.wait(0.1):
```

Comment:
Thanks Dave (as always!) for your detailed explanation. When I was looking at this code late last night, I hadn't realised the subtlety of there actually being _two_ nested while loops. And I still haven't quite got my head around there being so many different Event objects... ;)
But don't worry, I'm sure I'll get there eventually, so please focus on the code rather than taking time out to explain things to threading novices!